Can add an enum constant with lexical preserving printer

When I run the JavaParser it removes all empty lines within a method even if I do nothing to the CompilationUnit (some of them are there for a better overview and when I just add an EnumConstant I don't want half of the lines of the file to be changed...)
Is there any possibility to just modify what was added/removed to the CompilationUnit?

btw. Thanks for this tool its really useful!

Hi @matozoid  thx for your fast response!!

I tried the lexical preserving printer and it worked like you said, thank you! :)
... but it brought 2 little new problems:
When using the lexical preserving printer...

1. ...the JavaParser fails at adding EnumConstants when the enum ends with ';' (and no following code) [see Exception] - this isn't a problem without the lexical preserving printer
2. ...a switch statement is added in the same line as the existing ones. Makes sense to me since I'm not using the PrettyPrinter anymore but maybe there is a possibility to use the PrettyPrinter just for added Entries?

Hope you could help me again :) I tried to solve this one my own with javadoc but I did not get far...
.
.
.
_Caused by: java.lang.UnsupportedOperationException: Csm token token(}) NodeText TOKEN ";"   <102>   (line 166,col 40)-(line 166,col 40)
	at com.github.javaparser.printer.lexicalpreservation.Difference.applyKeptDiffElement(Difference.java:471)
	at com.github.javaparser.printer.lexicalpreservation.Difference.apply(Difference.java:142)
	at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter$Observer.concreteListChange(LexicalPreservingPrinter.java:274)
	at com.github.javaparser.ast.observer.PropagatingAstObserver.listChange(PropagatingAstObserver.java:72)
	at com.github.javaparser.ast.NodeList.lambda$notifyElementAdded$1(NodeList.java:466)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at com.github.javaparser.ast.NodeList.notifyElementAdded(NodeList.java:466)
	at com.github.javaparser.ast.NodeList.add(NodeList.java:70)
	at com.github.javaparser.ast.body.EnumDeclaration.addEnumConstant(EnumDeclaration.java:146)_

--- a/javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/PrettyPrinterIssue2351Test.java
+++ b/javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/PrettyPrinterIssue2351Test.java
@@ -1,0 +1,27 @@
+package com.github.javaparser.printer.lexicalpreservation;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.Optional;
+
+import org.junit.jupiter.api.Test;
+
+import com.github.javaparser.ast.body.EnumDeclaration;
+
+// manage java.lang.UnsupportedOperationException: Csm token token(}) NodeText TOKEN ";"   <102>   (line 1,col 39)-(line 1,col 39)
+class PrettyPrinterIssue2351Test extends AbstractLexicalPreservingTest  {
+
+    @Test
+    void printingEnum2() {
+        String def2 = 
+                "public enum RandomEnum {" + 
+                " TYPE_1;" +
+                "}";
+        considerCode(def2);
+        Optional<EnumDeclaration> decl = cu.findFirst(EnumDeclaration.class);
+        if (decl.isPresent()) decl.get().addEnumConstant("SOMETHING");
+        System.out.println(LexicalPreservingPrinter.print(cu));
+        assertTrue(decl.get().getEntries().size() == 2);
+    }
+
+}

--- a/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/ChildTextElement.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/ChildTextElement.java
@@ -99,6 +99,11 @@ class ChildTextElement extends TextElement {
     public boolean isComment() {
         return child instanceof Comment;
     }
+    
+    @Override
+    public boolean isSeparator() {
+        return false;
+    }
 
     @Override
     public boolean isChildOfClass(Class<? extends Node> nodeClass) {

--- a/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/Difference.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/Difference.java
@@ -463,6 +463,11 @@ public class Difference {
             } else if (kept.isNewLine() && originalTextToken.isSpaceOrTab()) {
                 originalIndex++;
                 diffIndex++;
+             // case where originalTextToken is a separator like ";" and 
+             // kept is not a new line or whitespace for example "}"
+             // see issue 2351
+            }  else if (!kept.isNewLine() && originalTextToken.isSeparator()) {
+                originalIndex++;
             } else if (kept.isWhiteSpaceOrComment()) {
                 diffIndex++;
             } else if (originalTextToken.isWhiteSpaceOrComment()) {

--- a/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TextElement.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TextElement.java
@@ -53,6 +53,8 @@ public abstract class TextElement implements TextElementMatcher {
     public abstract boolean isNewline();
 
     public abstract boolean isComment();
+    
+    public abstract boolean isSeparator();
 
     public final boolean isWhiteSpaceOrComment() {
         return isWhiteSpace() || isComment();

--- a/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java
@@ -104,6 +104,11 @@ class TokenTextElement extends TextElement {
     public boolean isComment() {
         return token.getCategory().isComment();
     }
+    
+    @Override
+    public boolean isSeparator() {
+        return token.getCategory().isSeparator();
+    }
 
     @Override
     public boolean isNewline() {

